src/explore-education-statistics-common/src/utils/url/applyRedirectRules.ts

```diff
@@ -0,0 +1,93 @@
+import redirectService, {
+  RedirectType,
+  redirectPathStarts,
+} from '@common/services/redirectService';
+
+const prodPublicUrl = 'https://explore-education-statistics.service.gov.uk/';
+
+export default async function applyRedirectRules(url: string): Promise<string> {
+  let newUrl: string = url;
+
+  newUrl = await noTrailingSlashUnlessOnlyHost(newUrl);
+  newUrl = await pathMustNotEndInSlash1000(newUrl);
+  newUrl = await publicSiteHostMustNotStartWithWww(newUrl);
+  newUrl = await mustNotTriggerAContentApiRedirect(newUrl);
```

Comment:
IMO, this approach is potentially over-complicating things by trying to be functional. 

I'm not really sure it's worth the effort as we don't typically write code like this, and it's less performant as it has to parse into a `URL` object multiple times and is using regexes over the entire URL in multiple places.

Would suggest that we just construct the `URL` object once and potentially just simplifying this back to some `if` statements (with comments) which modify the `url` i.e.

```ts
const newUrl = new URL(url);

// Do a check
if (blah) {
  newUrl.pathname = ...;
}

// Do another check
if (blah) {
  newUrl.hostname = ...;
}

// Using functions is still fine for more
// contained pieces of logic.
await callAFunction(newUrl);

return newUrl.toString();
```